File: code_snippets/ts/src/use_cases/role_updater.ts
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate Examples for the Node.js/TypeScript SDK,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/blob/main/LICENSE
 */

import * as restate from "@restatedev/restate-sdk";
import {Context} from "@restatedev/restate-sdk";

// <start_here>
// <mark_1>
const roleUpdater = restate.object({
    // </mark_1>
    name: "roleUpdate",
    handlers: {
        // <mark_1>
        // <mark_2>
        update: async function (ctx: Context, update: UpdateRequest) {
            // </mark_2>
            // </mark_1>
            const { userId, role, permissions: permissions } = update;

            // <mark_3>
            const previousRole = await ctx.run(() => getCurrentRole(userId));
            await ctx.run(() => applyUserRole(userId, role));
            // </mark_3>

            const previousPermissions: Permission[] = [];
            for (const permission of permissions) {
                // <mark_4>
                try {
                    // <mark_3>
                    const previous = await ctx.run(() =>
                        applyPermission(userId, permission));
                    // </mark_3>
                    previousPermissions.push(previous);
                } catch (err) {
                    if (err instanceof restate.TerminalError) {
                        await rollback(ctx, userId, previousRole, previousPermissions);
                    }
                    throw err;
                }
                // </mark_4>
            }
        }
    },
})
// <end_here>

async function rollback(
    ctx: restate.Context,
    userId: string,
    role: UserRole,
    permissions: Permission[]
) {
    console.log(">>> !!! ROLLING BACK CHANGES !!! <<<");
    for (const prev of permissions.reverse()) {
        await ctx.run(() => applyPermission(userId, prev));
    }
    await ctx.run(() => applyUserRole(userId, role));
}

export type UserRole = {
    roleKey: string;
    roleDescription: string;
};

export type Permission = {
    permissionKey: string;
    setting: string;
};

export type UpdateRequest = {
    userId: string;
    role: UserRole;
    permissions: Permission[];
};

export async function getCurrentRole(userId: string): Promise<UserRole> {
    // in this example, the previous role was always just 'viewer'
    return { roleKey: "viewer", roleDescription: "User that cannot do much" };
}

export async function applyUserRole(userId: string, userRole: UserRole): Promise<boolean> {
    console.log(`>>> Applied role ${userRole.roleKey} for user ${userId}`);
    return true;

}

export async function applyPermission(
    userId: string,
    permission: Permission
): Promise<Permission> {
    const { permissionKey, setting } = permission;
    return { permissionKey, setting: "blocked" };
}

File: code_snippets/go/usecases/microservices/roleupdater.go
package microservices

import (
	"context"

	restate "github.com/restatedev/sdk-go"
)

// <start_here>
// <mark_1>
type RoleUpdate struct{}

// <mark_2>
func (RoleUpdate) Update(ctx restate.Context, update UpdateRequest) error {
	// </mark_2>
	// </mark_1>

	// <mark_3>
	previousRole, err := restate.Run(ctx, func(ctx restate.RunContext) (UserRole, error) {
		return getCurrentRole(ctx, update.UserId)
	})
	if err != nil {
		return err
	}
	if _, err := restate.Run(ctx, func(ctx restate.RunContext) (bool, error) {
		return applyUserRole(ctx, update.UserId, update.Role)
	}); err != nil {
		return err
	}
	// </mark_3>

	var previousPermissions []Permission
	for _, permission := range update.Permissions {
		// <mark_4>
		// <mark_3>
		previous, err := restate.Run(ctx, func(ctx restate.RunContext) (Permission, error) {
			return applyPermission(ctx, update.UserId, permission)
		})
		if err != nil {
			if restate.IsTerminalError(err) {
				if err := rollback(ctx, update.UserId, previousRole, previousPermissions); err != nil {
					return err
				}
			}
			return err
		}
		// </mark_3>
		previousPermissions = append(previousPermissions, previous)
		// </mark_4>
	}

	return nil
}

// <end_here>

func rollback(
	ctx restate.Context,
	userId string,
	role UserRole,
	permissions []Permission,
) error {
	ctx.Log().Info(">>> !!! ROLLING BACK CHANGES !!! <<<")
	for i := len(permissions) - 1; i >= 0; i-- {
		if _, err := restate.Run(ctx, func(ctx restate.RunContext) (Permission, error) {
			return applyPermission(ctx, userId, permissions[i])
		}); err != nil {
			return err
		}
	}
	if _, err := restate.Run(ctx, func(ctx restate.RunContext) (bool, error) {
		return applyUserRole(ctx, userId, role)
	}); err != nil {
		return err
	}
	return nil
}

type UserRole struct {
	RoleKey         string
	RoleDescription string
}

type Permission struct {
	PermissionKey string
	Setting       string
}

type UpdateRequest struct {
	UserId      string
	Role        UserRole
	Permissions []Permission
}

func applyUserRole(
	ctx context.Context,
	userId string,
	userRole UserRole,
) (bool, error) {
	return true, nil
}

func applyPermission(
	ctx context.Context,
	userId string,
	permission Permission,
) (Permission, error) {
	return Permission{PermissionKey: permission.PermissionKey, Setting: "blocked"}, nil
}

func getCurrentRole(ctx context.Context, userId string) (UserRole, error) {
	// in this example, the previous role was always just 'viewer'
	return UserRole{RoleKey: "viewer", RoleDescription: "User that cannot do much"}, nil
}
